File: security/src/main/java/com/kmdev/security/UserService.java
package com.kmdev.security;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;
import static java.util.Objects.requireNonNull;

/**
 * User service class provides business functions to handle users.
 *
 * @author keuller.magalhaes at gmail.com
 */
public final class UserService {

    private List<User> userList = new ArrayList<>(2);

    UserService() {
        User user = User.build();
        user.username = "Dudu";
        user.password = "abc123";
        user.email = "dudu.magalhaes@gmail.com";
        user.role = "manager";
        userList.add(user);

        user = User.build();
        user.username = "abdoral";
        user.password = "Yt4d20V";
        user.email = "abdoral.gusmao@outlook.com";
        user.role = "admin";
        userList.add(user);
    }

    final Function<JsonObject, JsonObject> authenticate = (json) -> {
        requireNonNull(json);
        final String usr = json.getString("username");
        final String pwd = json.getString("password");
        Optional<User> user = userList.stream().filter(
            item -> item.username.equalsIgnoreCase(usr) && item.password.equals(pwd)
        ).findFirst();
        return (user.isPresent() ? user.get().toJson() : new JsonObject().put("message", "User not found."));
    };

    final Function<JsonObject, JsonObject> createUser = (json) -> {
        requireNonNull(json);
        final User user = User.build().fromJson(json);
        if ("".equals(user.email) || "".equals(user.role)) return new JsonObject().put("message", "Invalid user data.");
        userList.add(user);
        return new JsonObject().put("message", "User has been created.");
    };

    final Function<String, JsonObject> findByEmail = (mail) -> {
        requireNonNull(mail);
        Optional<User> user = userList.stream().filter(item -> item.email.equalsIgnoreCase(mail)).findFirst();
        return (user.isPresent() ? user.get().toJson() : new JsonObject().put("message", "User not found."));
    };

    final Supplier<JsonObject> findAll = () -> {
        final JsonArray users = new JsonArray();
        userList.stream().forEach(user -> users.add(user.toJson()));
        return new JsonObject().put("users", users);
    };

}

File: security/src/main/kotlin/com/kmdev/security/UserService.kt
package com.kmdev.security

import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import java.util.Objects.requireNonNull

internal val users = arrayListOf<User>()

fun handleCommand(cmd: String, data: JsonObject): JsonObject {
    return when(cmd) {
        "auth" -> authenticate(data.getString("username"), data.getString("password"))
        "create" -> createUser(data)
        "findAll" -> findAll()
        else -> JsonObject().put("message", "Invalid command.")
    }
}

fun authenticate(user: String, pass: String): JsonObject {
    if (user.isEmpty() || pass.isEmpty()) return JsonObject().put("message", "Invalid credentials.")
    return try {
        users.asSequence().filter { it.username.equals(user, true) && it.password == pass }
                .first()
                .toJson()
    } catch (ex: NoSuchElementException) {
        JsonObject().put("message", "User not found.")
    }
}

fun createUser(user: JsonObject): JsonObject {
    requireNonNull(user)
    val bean = User().fromJson(user)
    if (bean.username.isEmpty() || bean.email.isEmpty() || bean.role.isEmpty())
        return JsonObject().put("message", "Invalid user data.")
    users.add(bean)
    return JsonObject().put("message", "User has been created.")
}

fun findAll(): JsonObject {
    val result = JsonArray()
    users.forEach { result.add(it.toJson()) }
    return JsonObject().put("users", result)
}
